lib/ddtrace/trace_operation.rb

```diff
@@ -0,0 +1,400 @@
+      class TraceFinished < Datadog::Event
+        def initialize
+          super(:trace_finished)
+        end
+      end
+    end
+
+    private
+
+    attr_reader \
+      :events,
+      :root_span
+
+    def activate_span!(span_op)
+      @active_span = span_op
+      set_root_span!(span_op) unless root_span
+    end
+
+    def deactivate_span!(span_op)
+      # Set parent to closest unfinished ancestor span.
+      # Prevents wrong span from being set as the active span
+      # when spans finish out of order.
+      span_op = span_op.send(:parent) while !span_op.nil? && span_op.finished?
+      @active_span = span_op
+
+      # Set finished, to signal root span has completed.
+      @finished = true if span_op.nil?
+
+      span_op
+    end
+
+    def start_span(span_op)
+      begin
+        activate_span!(span_op)
+
+        # Update active span count
+        @active_span_count += 1
+
+        # Publish :span_before_start event
+        events.span_before_start.publish(span_op, self)
+      rescue StandardError => e
+        Datadog.logger.debug("Error starting span on trace: #{e} Backtrace: #{e.backtrace.first(3)}")
+      end
+    end
+
+    def finish_span(span, span_op, parent)
+      begin
+        # Save finished span & root span
+        @spans << span unless span.nil?
+
+        # Deactivate the span, re-activate parent.
+        deactivate_span!(parent)
+
+        # Update active span count
+        @active_span_count -= 1
+
+        # Publish :span_finished event
+        events.span_finished.publish(span, self)
+
+        # Publish :trace_finished event
+        events.trace_finished.publish(self) if finished?
+      rescue StandardError => e
+        Datadog.logger.debug("Error finishing span on trace: #{e} Backtrace: #{e.backtrace.first(3)}")
+        Datadog.health_metrics.error_span_finish(1, tags: ["error:#{e.class.name}"])
+      end
+    end
+
+    # Track the root span
+    def set_root_span!(span)
+      return if span.nil? || root_span
+
+      @root_span = span
+
+      # Auto populate these attributes if
+      # they haven't been set yet.
+      @name ||= span.name
+      @resource ||= span.resource
+      @service ||= span.service
+    end
+
+    def build_span_options(
+      context: nil,
+      events: nil,
+      on_error: nil,
+      resource: nil,
+      service: nil,
+      start_time: nil,
+      tags: nil,
+      type: nil
+    )
+      # Add default options
+      options = { trace_id: @id }
+      options[:on_error] = on_error unless on_error.nil?
+      options[:resource] = resource unless resource.nil?
+      options[:service] = service unless service.nil?
+      options[:start_time] = start_time unless start_time.nil?
+      options[:tags] = tags unless tags.nil?
+      options[:type] = type unless type.nil?
+      # TODO: Do we need to pass this through for backwards compatibility?
+      #       Should remove this when we can.
+      options[:context] = context
```

Comment:
I think we can keep this for `1.0`, adding a deprecation warning to `span.context` invocations, and later remove it in `2.0`.

But, if keeping `span.context` takes a lot of work in this PR, then I'm in favour of removing it.

Reply:
It does, and I actually eliminated all the usage of it, with the exception of a few internals (namely `ManualTracing` which I'd like to remove, and OpenTracer which we'd like to deprecate.)